### PR TITLE
reimplement `eliminate_outer_join`

### LINKED ISSUE #4270: reimplement `reduce_outer_join`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #4270
Part of #4267

reimplement `eliminate_outer_join` to avoid to use `global-state`.

In new implementation, we just depends only on the shape of the subtree.

# Rationale for this change

We don't have to worry about the filter not being above the Join.

Because this rule will cooperate with `filter_push_down`. If a predicate which can eliminate outer exist, it will be pushdown and stop above an outer-join.

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/eliminate_outer_join.rs [hunk 1]
```diff
@@ -39,191 +38,107 @@ impl ReduceOuterJoin {
...
 
+/// Attempt to eliminate outer joins to inner joins.
+/// for query: select ... from a left join b on ... where b.xx = 100;
+/// if b.xx is null, and b.xx = 100 returns false, filtered those null rows.
+/// Therefore, there is no need to produce null rows for output, we can use
+/// inner join instead of left join.
+///
+/// Generally, an outer join can be eliminated to inner join if equals from where
+/// return false while any inputs are null and columns of those equals are come from
+/// nullable side of outer join.
```
Comment (HaoYang670) [inline datafusion/optimizer/src/eliminate_outer_join.rs:49]: We can simplify the docs to just illustrate that " We will replace the outer join by the inner join if the `null` rows in the nullable side will be filtered out.

File: datafusion/optimizer/src/eliminate_outer_join.rs [hunk 2]
```diff
@@ -39,191 +38,107 @@ impl ReduceOuterJoin {
     }
 }
 
+/// Attempt to eliminate outer joins to inner joins.
```
[the remark this answers is not in the document]
Comment (jackwener) [reply datafusion/optimizer/src/eliminate_outer_join.rs:41]: You can see `fn eliminate_outer()`
Your example, this rule can do.
We can eliminate Full -> ROJ/LOJ/Inner

File: datafusion/optimizer/src/eliminate_outer_join.rs [hunk 3]
```diff
@@ -15,235 +15,141 @@
...
-                Ok(LogicalPlan::Filter(Filter::try_new(
-                    filter.predicate().clone(),
-                    Arc::new(reduce_outer_join(
-                        _optimizer,
-                        filter.input(),
-                        nonnullable_cols,
-                        _optimizer_config,
-                    )?),
-                )?))
+pub fn eliminate_outer(
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_outer_join.rs:104]: Thank you for eliminating the duplication

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Filed a follow on PR to add some docstrings: https://github.com/apache/arrow-datafusion/pull/4343

**Pull request merged**